src/platform/web/ui/session/room/timeline/TextMessageView.js

```diff
@@ -21,28 +21,30 @@ import {renderMessage} from "./common.js";
 
 export class TextMessageView extends TemplateView {
     render(t, vm) {
-        const bodyView = t.mapView(vm => vm.text, text => new BodyView(text));
+        const bodyView = t.mapView(vm => vm.messageFormat, messageFormat => new BodyView(messageFormat));
         return renderMessage(t, vm,
             [t.p([bodyView, t.time({className: {hidden: !vm.date}}, vm.date + " " + vm.time)])]
         );
     }
 }
 
 class BodyView extends StaticView {
+
+    get _formatFunction() {
+        return {
+            text: (param) => text(param.obj.text),
+            link: (param) => param.t.a({ href: param.obj.url, target: "_blank", rel: "noopener" }, [text(param.obj.text)]),
+            newline: (param) => param.t.br()
+        };
+    }
+
     render(t, value) {
-        const lines = (value || "").split("\n");
-        if (lines.length === 1) {
-            return text(lines[0]);
-        }
-        const elements = [];
-        for (const line of lines) {
-            if (elements.length) {
-                elements.push(t.br());
-            }
-            if (line.length) {
-                elements.push(t.span(line));
-            }
+        const children = [];
+        for (const m of value) {
+            const f = this._formatFunction[m.type];
+            const element = f({ obj: m, t: t });
```

Comment:
perhaps don't create an extra object here, but just pass `m` and `t` as separate arguments? Otherwise looks great!